x11/urt/man/man1/ppmtorle.1
.\" Copyright (c) 1990, Minnesota Supercomputer Center, Inc.
.TH PPMTORLE 1 "July 30, 1990" 1
.SH NAME
ppmtorle \- convert a PBMPLUS/ppm image file into an RLE image file.
.SH SYNOPSIS
.B ppmtorle
[
.B \-h
] [
.B \-v
] [
.B \-a
] [
.BI \-o " outfile"
] [
.I infile.ppm
]
.SH DESCRIPTION
This program converts PBMPLUS full-color (ppm) image files into Utah
.IR RLE (5)
image files.  PBMPLUS/ppm image files contain the image dimensions and image
data in the form of RGB triplets.  When converting to an RLE file, the alpha
channel may be optionally computed.  The origins of PBMPLUS and Utah RLE files
are  in the upper left and lower left corners respectively.  Therefore, it is
common practice to send the output of this program through the "rleflip"
utility -- see examples below.
.PP
The RLE file will contain a "true color" image
(24 bits).  These RLE files may then be viewed using any RLE image viewer.  When
they are displayed on an 8 bit display, the image must be dithered.  In order
to produce a better looking image (on 8 bit displays), it is recommended that
the image be quantizing (to 8 bit mapped color) prior to its display.  This may
be done by piping the output of this program into the Utah
.IR mcut (1)
or
.IR rlequant (1)
utilities.
An example of this is shown later.
.PP
.SH OPTIONS
.TP
.B \-v
This option will cause ppmtorle to operate in verbose mode.  The header
information is written to "stderr".  Actually, there is not much header
information stored in a PBMPLUS file, so this information is minimal.
.TP
.B \-h
This option allows the header of the PBMPLUS image to be dumped to "stderr"
without converting the file.  It is equivalent to using the \-v option except
that no file conversion takes place.
.TP
.B \-m
This option will cause ppmtorle to use the RGB data to compute an alpha
channel in the resulting RLE file.  For any non-zero RGB data, the alpha
channel will contain a value of 255.  The resulting RLE image file will
contain three color channels and an alpha channel.
.TP
.BI \-o " outfile"
If specified, the output will be written to this file.  If 
.I outfile
is "\-", or if it is not specified, the output will be written to the
standard output stream.
.TP
.I infile.ppm
The name of the PBMPLUS image data file to be converted.  This file must end
in ".ppm".  However, it is not necessary to supply the ".ppm" extension as it
will be added to the supplied name if not already there.
.SH EXAMPLES
.TP
ppmtorle \-v test.ppm \-o test.rle
While running in verbose mode, convert test.ppm to RLE format and store
resulting data in test.rle.
.TP
ppmtorle test | rleflip \-v | mcut >test.rle
Convert test.ppm to RLE format, flip if vertically, and convert to 8 bit
mapped color before storing data in test.rle
.TP
ppmtorle \-h test
Dump the header information of the PBMPLUS file called test.ppm.
.SH SEE ALSO
.IR mcut (1),
.IR pgmtorle (1),
.IR rlequant (1),
.IR rletoppm (1),
.IR urt (1),
.IR RLE (5).
.SH AUTHOR
.br
Wesley C. Barris
.br
Army High Performance Computing Research Center (AHPCRC)
.br
Minnesota Supercomputer Center, Inc.


x11/urt/cnv/ppmtorle.c
/*
 * This software is copyrighted as noted below.  It may be freely copied,
 * modified, and redistributed, provided that the copyright notice is
 * preserved on all copies.
 *
 * There is no warranty or other guarantee of fitness for this software,
 * it is provided solely "as is".  Bug reports or fixes may be sent
 * to the author, who may or may not act on them as he desires.
 *
 * You may not include this software in a program or other software product
 * without supplying the source, or without informing the end-user that the
 * source is available for no extra charge.
 *
 * If you modify this software, you should include a notice giving the
 * name of the person performing the modification, the date of modification,
 * and the reason for such modification.
 */
/*
 * ppmtorle - A program which will convert pbmplus/ppm images
 *            to Utah's "rle" image format.
 *
 * Author:      Wesley C. Barris
 *              AHPCRC
 *              Minnesota Supercomputer Center, Inc.
 * Date:        Fri July 20 1990
 * Copyright (c) Wesley C. Barris
 */
/*-----------------------------------------------------
 * System includes.
 */
#include <stdio.h>
#include <ppm.h>
#include <rle.h>

#ifdef USE_STDLIB_H
#include <stdlib.h>
#else

#ifdef USE_STRING_H
#include <string.h>
#else
#include <strings.h>
#define strrchr rindex
#endif

#ifdef VOID_STAR
extern void *malloc();
#else
extern char *malloc();
#endif
extern void free();

#endif /* USE_STDLIB_H */

#define VPRINTF if (verbose || header) fprintf

typedef unsigned char U_CHAR;
/*
 * Global variables.
 */
FILE	*fp;
int	format;
int	width, height;
int	verbose = 0, header = 0, do_alpha = 0;
gray	maxval;
/*-----------------------------------------------------------------------------
 *                                        Read the Wavefront image file header.
 */
void read_ppm_header()
{
   ppm_readppminit(fp, &width, &height, &maxval, &format);
   VPRINTF(stderr, "Image type: 24 bit true color\n");
   VPRINTF(stderr, "Full image: %dx%d\n", width, height);
   VPRINTF(stderr, "Maxval:     %d\n", maxval);
   if (do_alpha)
      VPRINTF(stderr, "Computing alpha channel...\n");
}
/*-----------------------------------------------------------------------------
 *                                             Write the rle image file header.
 */
void write_rle_header()
{
   rle_dflt_hdr.xmin    = 0;
   rle_dflt_hdr.xmax    = width-1;
   rle_dflt_hdr.ymin    = 0;
   rle_dflt_hdr.ymax    = height-1;
   rle_dflt_hdr.ncolors = 3;
   rle_dflt_hdr.background = 0;
   if (do_alpha) {
      rle_dflt_hdr.alpha = 1;
      RLE_SET_BIT(rle_dflt_hdr, RLE_ALPHA);
      }
   RLE_SET_BIT(rle_dflt_hdr, RLE_RED);
   RLE_SET_BIT(rle_dflt_hdr, RLE_GREEN);
   RLE_SET_BIT(rle_dflt_hdr, RLE_BLUE);
   rle_put_setup(&rle_dflt_hdr);
}
/*-----------------------------------------------------------------------------
 *                                      Write the rle data portion of the file.
 */
void write_rle_data()
{
   register int		x;
   register int		scan;
   register pixel	*pixelrow, *pP;
   rle_pixel		**scanline;
/*
 * Allocate some memory.
 */
   pixelrow = ppm_allocrow(width);
   if (rle_row_alloc(&rle_dflt_hdr, &scanline)) {
      fprintf(stderr, "Unable to malloc space for pixels\n");
      exit(-1);
      }
/*
 * Loop through the ppm files image window, write blank lines outside
 * active window.
 */
   for (scan = 0; scan < height; scan++) {
      ppm_readppmrow(fp, pixelrow, width, maxval, format);
      for (x = 0, pP = pixelrow; x < width; x++, pP++) {
         scanline[RLE_RED][x]   = PPM_GETR(*pP);
         scanline[RLE_GREEN][x] = PPM_GETG(*pP);
         scanline[RLE_BLUE][x]  = PPM_GETB(*pP);
         if (do_alpha) {
            scanline[RLE_ALPHA][x] = (scanline[RLE_RED][x] ||
                                     scanline[RLE_GREEN][x] ||
                                     scanline[RLE_BLUE][x] ? 255 : 0);
            }
         }
      rle_putrow(scanline, width, &rle_dflt_hdr);
      }
   VPRINTF(stderr, "Done -- write oef to RLE data.\n");
   rle_puteof(&rle_dflt_hdr);
}
/*-----------------------------------------------------------------------------
 *                      Convert an Wavefront image file into an rle image file.
 */
int
main(argc, argv)
int argc;
char **argv;
{
   char		*periodP, *ppmname = NULL, *outname = NULL;
   static char	filename[BUFSIZ];
   int		oflag;
/*
 * Get those options.
 */
   if (!scanargs(argc,argv,
       "% v%- h%- a%- o%-outfile!s infile.ppm%s",
       &verbose,
       &header,
       &do_alpha,
       &oflag, &outname,
       &ppmname))
      exit(-1);
/*
 * Open the file.
 */
   if (ppmname == NULL) {
      strcpy(filename, "stdin");
      fp = stdin;
      }
   else {
      periodP = strrchr(ppmname, '.');
      strcpy(filename, ppmname);
      if (periodP) {
         if (strcmp(periodP, ".ppm")) /* does not end in ppm */
            strcat(filename, ".ppm");
         }
      else				/* no ext -- add one */
         strcat(filename, ".ppm");
      if (!(fp = fopen(filename, "r"))) {
         fprintf(stderr, "Cannot open %s for reading.\n", filename);
         exit(-1);
         }
      }
/*
 * Read the Wavefront file file header.
 */
   read_ppm_header();
   if (header)
      exit(0);
/*
 * Write the rle file header.
 */
   rle_dflt_hdr.rle_file = rle_open_f( cmd_name( argv ), outname, "w" );
   rle_addhist(argv, (rle_hdr *)NULL, &rle_dflt_hdr);
   write_rle_header();
/*
 * Write the rle file data.
 */
   write_rle_data();
   fclose(fp);

   return 0;
}
